Add ci.py helper script to allow parametrizing the pants_version

Eric-Arellano opened the pull request:
### Problem
The `./pants` script claims to support running with both a pinned `pants_version` in the `pants.ini` and with the `pants_version` unspecified.

We currently do not test this unspecified branch, which we should be doing to guarantee that functionality always works.

Further, this is pre-work for https://github.com/pantsbuild/setup/pull/32, which will add `--python-version` as another parameter for `ci.py` to allow an unspecified Python version vs. `pants.ini` specifying 2.7 or 3.6 (FYI resulting [`.travis.yml`](https://github.com/pantsbuild/setup/pull/32/files#diff-354f30a63fb0907d4ad57269548329e3)).

### Solution
Introduce a new `ci.py` helper script that uses argparse to have the parameter `--pants-version {unspecified,config}`.

#### Requires adding Python 3 to CI shards
Because this script uses modern Python 3—i.e. enums, f-strings, type hints, keyword-only args, and `subprocess.run()`—it requires Python 3.5+ to run.

Even though we can't yet run `./pants` with Python 3 until https://github.com/pantsbuild/setup/pull/32 lands, this PR does some of the pre-work in adding Python 3 to each of the shards so that we can run this script. This is valid to do because every single shard is going to run the tests with both Py27 and Py36—unlike Pants where we have dedicated shards for each interpreter—so there is only benefit in ensuring each shard has Py27 and Py36.

#### Also updates test's pinned `pants_version`

Eric-Arellano committed:
Add ci.py to CI to allow parameterizing pants_version and python_version

Eric-Arellano committed:
Remove PythonVersion and fix --pants-version unspecified

--pants-version unspecified would fail because of a bad config, that the contrib module would refer to pants_version even though that line was deleted.

Eric-Arellano committed:
Try to install Py36 on each shard

Also modernize .travis.yml to use all the cool goodies we've learned in Pants's CI like anchors and the brew addon package!

Eric-Arellano committed:
Tweak ci.py formatting

Eric-Arellano committed:
Add comment explaining why we rewrite pants.ini like this

Eric-Arellano committed:
Add  prefix to script. Oops

Eric-Arellano committed:
Merge branch 'gh-pages' of github.com:pantsbuild/setup into ci-py

Eric-Arellano committed:
Remove stray whitespace

Eric-Arellano committed:
Try to debug Trusty image using Py3.4

Eric-Arellano committed:
Remove unneeded anchors from OSX shards

Turns out Precise comes installed with 3.6 when requested! So we only need Pyenv for OSX, not Linux.

Eric-Arellano committed:
Test if OSX comes pre-installed with Python?

Eric-Arellano committed:
Maybe fix Trusty using Py3.4

Eric-Arellano committed:
Global python entry -> build matrix

See if moving back to osx_setup and linux_setup works.

Eric-Arellano committed:
Try pyenv approach

Path wrangling didn't work.

Eric-Arellano committed:
Set Pyenv for Trusty

Eric-Arellano committed:
Move constant to top of file

Think this is the style used in most Python projects and ours.

Eric-Arellano committed:
OSX requires language: general

Eric-Arellano committed:
Trusty Py34 is from system, not pyenv

Eric-Arellano committed:
Rename variables to makes less confusing what is pants_version

Eric-Arellano committed:
Update pants_version to 1.14.0

stuhood commented on ci.py:
```diff
@@ -0,0 +1,72 @@
+
+class PantsVersion(Enum):
+  unspecified = "unspecified"
+  pants_ini = "pants.ini"
```
Slightly confusing to have an enum choice that looks like a filename. Maybe just `config`?

stuhood reviewed (approved):
Looks good. Thanks!

Eric-Arellano committed:
Rename pants.ini cli arg to config

Good suggestion Stu!

Eric-Arellano committed:
Upgrade all the way to 1.15.0.dev3

Will lead to less of a diff on Friday when we go to 1.15.0.dev4. Plus now the pants daemon tests work on macOS for me!

Eric-Arellano committed:
Unify homebrew env entry

It's unified in Pants repo, so no good reason to split here.
